index.js

```diff
@@ -374,30 +375,73 @@ Pbf.prototype = {
     }
 };
 
-function readVarintRemainder(val, pbf) {
-    var buf = pbf.buf, b;
+function readVarintRemainder(l, u, p) {
+    var buf = p.buf,
+        h, b;
 
-    b = buf[pbf.pos++]; val += (b & 0x7f) * 0x10000000;         if (b < 0x80) return val;
-    b = buf[pbf.pos++]; val += (b & 0x7f) * 0x800000000;        if (b < 0x80) return val;
-    b = buf[pbf.pos++]; val += (b & 0x7f) * 0x40000000000;      if (b < 0x80) return val;
-    b = buf[pbf.pos++]; val += (b & 0x7f) * 0x2000000000000;    if (b < 0x80) return val;
-    b = buf[pbf.pos++]; val += (b & 0x7f) * 0x100000000000000;  if (b < 0x80) return val;
-    b = buf[pbf.pos++]; val += (b & 0x7f) * 0x8000000000000000; if (b < 0x80) return val;
+    b = buf[p.pos++]; h  = (b & 0x70) >> 4;  if (b < 0x80) return toNum(l, h, u);
+    b = buf[p.pos++]; h |= (b & 0x7f) << 3;  if (b < 0x80) return toNum(l, h, u);
+    b = buf[p.pos++]; h |= (b & 0x7f) << 10; if (b < 0x80) return toNum(l, h, u);
+    b = buf[p.pos++]; h |= (b & 0x7f) << 17; if (b < 0x80) return toNum(l, h, u);
+    b = buf[p.pos++]; h |= (b & 0x7f) << 24; if (b < 0x80) return toNum(l, h, u);
+    b = buf[p.pos++]; h |= (b & 0x01) << 31; if (b < 0x80) return toNum(l, h, u);
 
     throw new Error('Expected varint not more than 10 bytes');
 }
 
+function toNum(low, high, isUnsigned) {
+    if (isUnsigned) {
+        return ((high >>> 0) * 0x100000000) + (low >>> 0);
+    }
+
+    return high * 0x100000000 + (low >>> 0);
+}
+
 function writeBigVarint(val, pbf) {
+    var low, high;
+
+    if (val >= 0) {
+        low  = (val % 0x100000000) | 0;
+        high = (val / 0x100000000) | 0;
+    } else {
+        low  = ~(-val % 0x100000000);
+        high = ~(-val / 0x100000000);
+
+        if (low ^ 0xffffffff) {
+            low = (low + 1) | 0;
+        } else {
+            low = 0;
+            high = (high + 1) | 0;
+        }
+    }
+
+    if (toNum(low, high, true) < Math.abs(val)) {
```

Comment:
Narrowed it down to this line causing the performance issue. Changing this to the following fixes the speed regression.

```js
    if (val >= 0x8000000000000000 || val < -0x8000000000000000) {
        throw new Error('Given varint doesn\'t fit into 10 bytes');
    }
```

```
decode vector tile with pbf x 920 ops/sec ±1.04% (89 runs sampled)
encode vector tile with pbf x 613 ops/sec ±1.11% (83 runs sampled
```